core/src/client/mod.rs

```diff
@@ -592,6 +655,61 @@ impl<'a, R> IntoIterator for BatchResponse<'a, R> {
 	}
 }
 
+#[derive(thiserror::Error, Debug)]
+enum TrySubscriptionSendError {
+	#[error("The subscription is closed")]
+	Closed,
+	#[error("A subscription message was dropped")]
+	TooSlow(JsonValue),
+}
+
+#[derive(Debug)]
+pub(crate) struct SubscriptionSender {
+	inner: mpsc::Sender<JsonValue>,
+	lagged: SubscriptionLagged,
+}
+
+impl SubscriptionSender {
+	fn send(&self, msg: JsonValue) -> Result<(), TrySubscriptionSendError> {
+		match self.inner.try_send(msg) {
+			Ok(_) => Ok(()),
+			Err(TrySendError::Closed(_)) => Err(TrySubscriptionSendError::Closed),
+			Err(TrySendError::Full(m)) => {
+				self.lagged.set_missed();
+				Err(TrySubscriptionSendError::TooSlow(m))
+			}
+		}
+	}
+}
+
+#[derive(Debug)]
+pub(crate) struct SubscriptionReceiver {
+	inner: mpsc::Receiver<JsonValue>,
+	lagged: SubscriptionLagged,
+}
+
+pub(crate) struct TooSlow;
+
+impl Stream for SubscriptionReceiver {
+	type Item = Result<JsonValue, TooSlow>;
+
+	fn poll_next(mut self: Pin<&mut Self>, cx: &mut task::Context<'_>) -> task::Poll<Option<Self::Item>> {
+		if self.lagged.has_missed_message() {
+			return Poll::Ready(Some(Err(TooSlow)));
```

Comment:
I wonder whether a potential issue here is that this "missed messages" thing being set to true doesn't wake up the task. It might be irrelevant though since, if messages are being missed, it is because this thing isn't being polled quickly enough already, so I'm inclined to think it's not an issue.

Reply:
`has_missed_message` will flip the flag and two reads will end up reading the underlying stream but it's possible that the stream will never be read if the sender/receiver at interleaved:

1. send (full)
2. read (too slow)
3. send (full)
4. read (too slow)
...

and so on but I find it unlikely to be an issue because it should be very unlikely...